tests/activity_api_reset_test.go

```diff
@@ -0,0 +1,328 @@
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.True(t, activityCompleted.Load() > 1)
+	}, 5*time.Second, 200*time.Millisecond)
+
+	resetRequest := &workflowservicepb.ResetActivityByIdRequest{
+		Namespace:  s.Namespace(),
+		WorkflowId: workflowRun.GetID(),
+		ActivityId: "activity-id",
+		NoWait:     true,
+	}
+	resp, err := s.FrontendClient().ResetActivityById(ctx, resetRequest)
+	s.NoError(err)
+	s.NotNil(resp)
+
+	activityReseted.Store(true)
+
+	// wait for activity to be running
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, enumspb.PENDING_ACTIVITY_STATE_STARTED, description.PendingActivities[0].State)
+		// also verify that the number of attempts was reset
+		assert.Equal(t, int32(1), description.PendingActivities[0].Attempt)
+	}, 5*time.Second, 100*time.Millisecond)
+
+	// let activity finish
+	activityCompleteSignal <- struct{}{}
+
+	// wait for workflow to complete
+	var out string
+	err = workflowRun.Get(ctx, &out)
+
+	s.NoError(err)
+}
+
+func (s *ActivityApiResetClientTestSuite) TestActivityResetApi_WithNoWait() {
+	// activity reset is called while activity is running, with NoWait=true to start new activity immediately
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	activityCompleteSignal1 := make(chan struct{})
+	activityCompleteSignal2 := make(chan struct{})
+	var activityReseted atomic.Bool
+
+	activityFunction := func() (string, error) {
+		if activityReseted.Load() == false {
+			activityErr := errors.New("bad-luck-please-retry")
+			s.WaitForChannel(ctx, activityCompleteSignal1)
+			return "", activityErr
+		}
+
+		s.WaitForChannel(ctx, activityCompleteSignal2)
+		return "done!", nil
+	}
+
+	workflowFn := s.makeWorkflowFunc(activityFunction)
+
+	s.Worker().RegisterWorkflow(workflowFn)
+	s.Worker().RegisterActivity(activityFunction)
+
+	wfId := testcore.RandomizeStr("wfid-" + s.T().Name())
+	workflowOptions := sdkclient.StartWorkflowOptions{
+		ID:        wfId,
+		TaskQueue: s.TaskQueue(),
+	}
+
+	workflowRun, err := s.SdkClient().ExecuteWorkflow(ctx, workflowOptions, workflowFn)
+	s.NoError(err)
+
+	// wait for activity to start
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, enumspb.PENDING_ACTIVITY_STATE_STARTED, description.PendingActivities[0].State)
+	}, 5*time.Second, 200*time.Millisecond)
+
+	activityReseted.Store(true)
```

Comment:
the previous test had this line _after_ the `ResetActivityByIdRequest` call; does the order matter?

Reply:
yes, in a subtle way